tests/test_cli_actions.py

```diff
@@ -0,0 +1,203 @@
+    def test_loop_until_finished_raises(self, res_msg, monkeypatch):
+        """Check that certain errors are raised and handled correctly by the
+        loop_until_finished function."""
+        with monkeypatch.context() as m:
+            m.setattr(
+                pennylane_orquestra.cli_actions, "workflow_details", lambda *args: "Some message1"
+            )
+            m.setattr(pennylane_orquestra.cli_actions, "workflow_results", lambda *args: res_msg)
+
+            # Check that indexing into the message raises an IndexError
+            # (this shows that it will be handled internally)
+            with pytest.raises(IndexError, match="list index out of range"):
+                location = res_msg[1].split()[1]
+
+            # Check that looping eventually times out
+            with pytest.raises(TimeoutError, match="were not obtained after"):
+                loop_until_finished("Some ID", timeout=1)
+
+    def test_loop_raises_workflow_fail(self, monkeypatch):
+        """Check that an error is raised if the workflow exeuction failed."""
+        with monkeypatch.context() as m:
+            status = "Status:              Failed\n"
+            result_message = "Some message2"
+
+            m.setattr(pennylane_orquestra.cli_actions, "workflow_details", lambda *args: status)
+            m.setattr(
+                pennylane_orquestra.cli_actions,
+                "workflow_results",
+                lambda *args: result_message,
+            )
+
+            # Check that looping raises an error if the workflow details
+            # contain a failed status
+            with pytest.raises(
+                ValueError, match=f"Something went wrong with executing the workflow. {status}"
+            ):
+                loop_until_finished("Some ID", timeout=1)
+
+    def test_valid_url(self, monkeypatch, tmpdir):
+        """Test that when receiving a valid url, data will be decoded and
+        returned."""
+        decoded_data = {"res": "Decoded Data"}
+        test_file = os.path.join(tmpdir, "workflow_result.json")
+        test_tar = os.path.join(tmpdir, "test.tgz")
+
+        with open(test_file, "w") as outfile:
+            json.dump(decoded_data, outfile)
+
+        tar = tarfile.open(test_tar, mode="w:gz")
+        tar.add(test_tar)
+        tar.close()
+
+        # Change to the test directory
+        os.chdir(tmpdir)
+        with monkeypatch.context() as m:
+            status = "Status:              Failed\n"
+            result_message = ["Some message2", "Some location"]
+
+            m.setattr(
+                pennylane_orquestra.cli_actions, "workflow_results", lambda *args: result_message
+            )
+            m.setattr(urllib.request, "urlopen", lambda arg: arg)
+            m.setattr(urllib.request, "urlretrieve", lambda *args, **kwargs: (test_tar,))
+            assert loop_until_finished("Some ID", timeout=1) == decoded_data
+
+    # Expected to fail if qe is unavailable
```

Comment:
When exactly do we expect this to fail? When the `qe` CLI is not available, or when the service is not accessible?

Reply:
Good question! It's when `qe` is unavailable. Added this so that the CI can pass.

Reply:
Weird. I ran it locally (where `qe` is available) and it still `xfail`d

Reply:
Hmm that sounds odd :thinking: Did it have `X` (upper case) or `x` (lower case)? For me locally it's `X` (passed but marked as `xfail`). Not logged in via `qe`.